plugins/network-device-injector/network-device-injector.go

```diff
@@ -0,0 +1,504 @@
+		dump("RunPodSandbox", "pod", pod)
+	}
+
+	// inject associated netdevices (based on received pod annotations) into the pod
+	// network namespace that will be attached to the pod's containers
+	netdevices, err := parseNetdevices(pod.Annotations)
+	if err != nil {
+		return err
+	}
+
+	if len(netdevices) == 0 {
+		return nil
+	}
+
+	// get the pod network namespace
+	var ns string
+	for _, namespace := range pod.Linux.GetNamespaces() {
+		if namespace.Type == "network" {
+			ns = namespace.Path
+			break
+		}
+	}
+
+	// Pods running on the host network namespace has this value empty
+	if ns == "" {
+		log.WithField("namespace", pod.GetNamespace()).WithField("name", pod.GetName).Info("Pod using host namespace, skipping ...")
+		return nil
+	}
+
+	// attach the network devices to the pod namespace
+	for _, n := range netdevices {
+		err = n.inject(ns)
+		if err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (p *plugin) StopPodSandbox(_ context.Context, pod *api.PodSandbox) error {
+	log.WithField("namespace", pod.GetNamespace()).WithField("name", pod.GetName).Debug("Stopped pod...")
+	if verbose {
+		dump("StopPodSandbox", "pod", pod)
+	}
+	// release associated devices of the netdevice to the Pod
+	netdevices, err := parseNetdevices(pod.Annotations)
+	if err != nil {
+		return err
+	}
+
+	if len(netdevices) == 0 {
+		return nil
+	}
+
+	// get the pod network namespace
+	var ns string
+	for _, namespace := range pod.Linux.GetNamespaces() {
+		if namespace.Type == "network" {
+			ns = namespace.Path
+			break
+		}
+	}
+	// TODO check host network namespace
+	if ns == "" {
+		return nil
+	}
+
+	// release the network devices from the pod namespace
+	for _, n := range netdevices {
+		err = n.release(ns)
+		if err != nil {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func parseNetdevices(annotations map[string]string) ([]netdevice, error) {
+	var (
+		key        string
+		annotation []byte
+		netdevices []netdevice
+	)
+
+	// look up effective device annotation and unmarshal devices
+	for _, key = range []string{
+		netdeviceKey + "/pod",
+		netdeviceKey,
+	} {
+		if value, ok := annotations[key]; ok {
+			annotation = []byte(value)
+			break
+		}
+	}
+
+	if annotation == nil {
+		return nil, nil
+	}
+
+	if err := yaml.Unmarshal(annotation, &netdevices); err != nil {
+		return nil, fmt.Errorf("invalid device annotation %q: %w", key, err)
+	}
+
+	// validate and default
+	for _, n := range netdevices {
+		if n.NewName == "" {
+			n.NewName = n.Name
+		}
+		if n.Address != "" {
+			ip := net.ParseIP(n.Address)
+			if ip == nil {
+				return nil, fmt.Errorf("error parsing address %s", n.Address)
+			}
+
+			if n.Prefix == 0 {
+				if ip.To4() == nil {
+					n.Prefix = 128
+				} else {
+					n.Prefix = 32
+				}
+			}
+		}
+
+	}
+	return netdevices, nil
+}
+
+// Dump one or more objects, with an optional global prefix and per-object tags.
+func dump(args ...interface{}) {
+	var (
+		prefix string
+		idx    int
+	)
+
+	if len(args)&0x1 == 1 {
```

Comment:
I'm confused why `&0x1` is needed here?

Reply:
copied pasted from the  device injector plugin, didn't notice it honestly